Add outer join breaking change (#2199)
ballerina-platform/ballerina-release: release-notes/swan-lake-release-notes/swan-lake-2201.3.0-release-note.md

int result = -9223372036854775808; // error: '9223372036854775808' is out of range for 'int'
```

> - Restricted the typed binding of the query expression outer join to be inferred as a nillable type. Earlier, the typed binding for outer join could be any type descriptor but now it's restricted to be `var` so that it will be inferred as a nillable type.

Reviewer: `Restricted the typed binding of the query expression outer join to be inferred as a nillable type.`
Shouldn't this be a heading?

Reviewer 2: it is too long for a heading. If we are making a heading out of it, we need to summarize it.

Author: Fixed

int result = -9223372036854775808; // error: '9223372036854775808' is out of range for 'int'
```

> #### Restricted the typed binding of the query expression outer join to `var`

Reviewer 3: Incomplete?

```suggestion
#### Restriction requiring the typed binding pattern of an outer join clause to use `var`
```

Author: Would the title length be fine @praneesha?

Reviewer 2: @MaryamZi - Can we summarize this more as it is a title?

Reviewer 3: Added some suggestions in https://github.com/ballerina-platform/ballerina-dev-website/pull/5812/files#r1023922325.

> Previously, the typed binding for outer join could be any type descriptor, but now it's restricted to be `var` so that it will be inferred as a nillable type.

Reviewer 3: ```suggestion
Previously, the typed binding pattern in an outer join clause was allowed to be any type descriptor, but now it has been restricted to be `var` so that an optional type (`T?`) will be inferred as the type.
```

```ballerina
type User record {|
    readonly int id;
    string name;
|};

type Login record {|
    int userId;
    string time;
|};

type Entry record {|
    string? name;
    string login;
|};

public function main() {
    Login[] logins = [
        {userId: 6789, time: "20:10:23"},
        {userId: 1234, time: "10:30:02"},
        {userId: 3987, time: "12:05:00"}
    ];

    table<User> key(id) users = table [
            {id: 1234, name: "Keith"},
            {id: 6789, name: "Anne"}
        ];

    Entry[] _ = from var login in logins
        outer join var user in users
        on login.userId equals user?.id
        select {name: user?.name, login: login.time};

>     // Now, the code below gives an error at the outer join typed binding.

Reviewer 3: ```suggestion
    // Now, the code below gives an error at the typed binding of the outer join clause.
```

    _ = from var login in logins

>         outer join User user in users  //error: outer join must be declared with 'var'

Reviewer 3: ```suggestion
        outer join User user in users  // error: outer join must be declared with 'var'
```

  - `createEnumDeclarationNode`
  - `createClassDefinitionNode`

- Restricted query expression outer join type-binding to be inferred as a nillable type 

>     Earlier, the type binding for outer join could be any type-descriptor but now its restricted to be 'var' so that it will be inferred as a nillable type.

Reviewer: ```suggestion
    Earlier, the type binding for outer join could be any type-descriptor but now it's restricted to be 'var' so that it will be inferred as a nillable type.
```

   
```ballerina
type User record {|
    readonly int id;
    string name;
|};

type Login record {|
    int userId;
    string time;
|};

public function main() {
    Login[] logins = [
        {userId: 6789, time: "20:10:23"},
        {userId: 1234, time: "10:30:02"},
        {userId: 3987, time: "12:05:00"}
    ];
    
    table<User> key(id) users = table [
            {id: 1234, name: "Keith"},
            {id: 6789, name: "Anne"}
        ];

    string[] _ = from var login in logins
        outer join var user in users 
        on login.userId equals user?.id

>         select user?.name + ":" + login.time;

Reviewer: This should be a error since this is `string?` + `string` right?

Author: Yes let me update

>     string[] _ = from var login in logins
>         outer join var user in users 
>         on login.userId equals user?.id
>         select user?.name + ":" + login.time;

Reviewer: ```suggestion
    Entry[] _ = from var login in logins
        outer join var user in users 
        on login.userId equals user?.id
        select {name: user?.name, login: login.time};
```
where 
```
type Entry record {|
    string? name;
    string login;
|};
```

  - `createEnumDeclarationNode`
  - `createClassDefinitionNode`

> - Restricted the type-binding for query expression outer join to be inferred as a nillable type 

Reviewer 2: ```suggestion
- Restricted the type binding for the query expression outer join to be inferred as a nillable type. 
```

> - Restricted the type-binding for query expression outer join to be inferred as a nillable type 
>     Earlier, the type binding for outer join could be any type-descriptor but now it's restricted to be 'var' so that it will be inferred as a nillable type.

Reviewer 4: ```suggestion
- Restricted the typed binding for query expression outer join to be inferred as a nillable type. Earlier, the type binding for outer join could be any type descriptor but now it's restricted to be `var` so that it will be inferred as a nillable type.
```

Reviewer 2: Not sure which is technically correct `typed binding` or `type binding`. Shall we check in the spec?

Author: ```suggestion
- Restricted the typed binding for query expression outer join to be inferred as a nillable type. Earlier, the typed binding for outer join could be any type descriptor but now it's restricted to be `var` so that it will be inferred as a nillable type.
```

>     Earlier, the type binding for outer join could be any type-descriptor but now it's restricted to be 'var' so that it will be inferred as a nillable type.

Reviewer 2: ```suggestion
    Earlier, the type binding for outer join could be any type descriptor but now it's restricted to 'var' so that it will be inferred as a nillable type.
```

   
```ballerina
type User record {|
    readonly int id;
    string name;
|};

type Login record {|
    int userId;
    string time;
|};

type Entry record {|
    string? name;
    string login;
|};

public function main() {
    Login[] logins = [
        {userId: 6789, time: "20:10:23"},
        {userId: 1234, time: "10:30:02"},
        {userId: 3987, time: "12:05:00"}
    ];

    table<User> key(id) users = table [
            {id: 1234, name: "Keith"},
            {id: 6789, name: "Anne"}
        ];

    Entry[] _ = from var login in logins
        outer join var user in users
        on login.userId equals user?.id
        select {name: user?.name, login: login.time};

>     // now, below gives an error at outer join type binding    

Reviewer 4: ```suggestion
    // Now, below gives an error at the outer join typed binding
```
